### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by RonalAlexis2**
BACK-14 Ticket Solution

This is a hands-on object-oriented programming exercise, done in C# in ID Visual Studio.
The concepts utilice is :

diagrams UML
Class.
methods
attributes
Inheritance
extends
Implements
Polymorphism.
Interface
Objects as parameter
model class
encapsulation

### CODE REVIEW

File: Ave.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        //sobre escribimos atacar de la clase padre
+        override public void atacar()
+        {
+            Console.WriteLine("el ave ataca");
+        }         
+
+        //creamos un nuevo metodo que reciba parametros
+        public string cantar( string trinar)
+        {
+            return trinar;
```
Comment (MauroDecodificado) [inline Ave.cs:21]: Please do something inside this method, do not just return the same variable

File: Colibri.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            Console.WriteLine("el colibri come");
+        }
+        //se crea comportmiento diferente
+        public void volar()
+        {
+            Console.WriteLine("el colibri vuela");
+        }
+
+        //se reciben los parametros de muerte y se calcula cuanto vivio
+        public string Morir(int nacimiento, int faleccimiento, string causa)
```
Comment (MauroDecodificado) [inline Colibri.cs]: Type error on faleccimiento please check it

File: Colibri.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            int calculo;
+            calculo = nacimiento - faleccimiento;
```
Comment (MauroDecodificado) [inline Colibri.cs]: You can put this as one single line 

int calculo =  nacimiento - fallecimiento;

File: Colibri.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            Console.WriteLine("La fecha de su muerte es: " + faleccimiento, "vivio "+ calculo, "años y la causa de la muerte es"+ causa);
+            string m=Console.ReadLine();
```
Comment (MauroDecodificado) [inline Colibri.cs]: Is there any reason why having this?

File: Colibri.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            return m;
+        }  
+        //se retorna el mismo string que recibimos
+        public string Reencarna(string revivio)
+        {
+            return revivio;
+        }
+        //creamos variable para guardar informacion del constructor        
+        ModeloAnimal guardarinfo;
```
Comment (MauroDecodificado) [inline Colibri.cs]: Global variables must be at the top of the class

File: Colibri.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        //constructor que no recibe parametros, pero llena info de ModeloAnimal
+        public Colibri()
+        { 
+            ModeloAnimal ocolibri=new ModeloAnimal();
```
Comment (MauroDecodificado) [inline Colibri.cs]: Error on name ocolibri, please check

File: Vaca.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+namespace EjercicioDos
+{
+    internal class Vaca : Mamifero, Muerte
+    {    //se sobrescribe el metodo de mamimero amamantar
+        public override void amamanta()
+        {
+            Console.WriteLine("la vaca da leche");
+        }
+        //logica para calcular murte y pares
+        public string Morir(int nacimiento, int faleccimiento, string causa)
```
Comment (MauroDecodificado) [inline Vaca.cs]: Error on name faleccimiento please check it

File: Vaca.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            int calculo = nacimiento - faleccimiento;
```
Comment (MauroDecodificado) [inline Vaca.cs]: You can put this in one single line 

int calculo = (nacimiento - fallecimiento) * 2

File: Vaca.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            calculo = calculo * 2;
+            string N;
```
Comment (MauroDecodificado) [inline Vaca.cs]: What is N? use a meaningful name

File: Vaca.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            {
+                N = "par";
+            }
+            else
+            {
+                N = "impar";
+            }
+
+            Console.WriteLine("La fecha de su muerte " +faleccimiento, "el numero es " + N, "vivio"+ calculo, "años y murio por "+ causa);
+            string m = Console.ReadLine();
```
Comment (MauroDecodificado) [inline Vaca.cs]: What is the reason to having this?

File: Vaca.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        }
+        //reencarna devuelve el mismo recibido pero concatenado
+        public string Reencarna(string revivio)
+        {
+            return revivio + "Tu puedes hacerlo!";
+        }
+        //constructor que no recibe parametros
+        public Vaca()
+        {
+            ModeloAnimal ovaca = new ModeloAnimal();
```
Comment (MauroDecodificado) [inline Vaca.cs:44]: You are forgetting assigning this to a global variable to keep it available in-memory object

**Review by MauroDecodificado — CHANGES_REQUESTED** [review]
Please check comments

File: Ave.cs (hunk 1 above)
Comment (RonalAlexis2) [reply Ave.cs:21]: has been corrected

File: Colibri.cs (hunk 2 above)
Comment (RonalAlexis2) [reply Colibri.cs]: has been corrected

File: Colibri.cs (hunk 4 above)
Comment (RonalAlexis2) [reply Colibri.cs]: tried in a variable to store the text and return it.

studying I realize that I can return the console.readline without the need for variables.

it has been fixed like this

File: Colibri.cs (hunk 3 above)
Comment (RonalAlexis2) [reply Colibri.cs]: has been corrected

File: Colibri.cs (hunk 5 above)
Comment (RonalAlexis2) [reply Colibri.cs]: has been corrected

File: Colibri.cs (hunk 6 above)
Comment (RonalAlexis2) [reply Colibri.cs]: has been corrected

File: Vaca.cs (hunk 7 above)
Comment (RonalAlexis2) [reply Vaca.cs]: has been corrected

File: Vaca.cs (hunk 8 above)
Comment (RonalAlexis2) [reply Vaca.cs]: has been corrected

File: Vaca.cs (hunk 9 above)
Comment (RonalAlexis2) [reply Vaca.cs]: a change has been made to the variable to be significant

File: Vaca.cs (hunk 10 above)
Comment (RonalAlexis2) [reply Vaca.cs]: tried in a variable to store the text and return it.

studying I realize that I can return the console.readline without the need for variables.

it has been fixed like this

File: Vaca.cs (hunk 11 above)
Comment (RonalAlexis2) [reply Vaca.cs:44]: has been corrected

File: Colibri.cs (hunk 4 above)
Comment (MauroDecodificado) [reply Colibri.cs]: You can have something like this

     

```
 public static string Morir(int nacimiento, int fallecimiento, string causa)
        {
            int calculo;
            calculo = nacimiento - fallecimiento;
            return "La fecha de su muerte es: " + fallecimiento.ToString() + " vivio " + calculo.ToString() + " años y la causa de la muerte es "+ causa;
        }
```

and then on program print the returned string

File: Colibri.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            Console.WriteLine("el colibri come");
+        }
+        //se crea comportmiento diferente
+        public void volar()
+        {
+            Console.WriteLine("el colibri vuela");
+        }
+
+        //se reciben los parametros de muerte y se calcula cuanto vivio
+        public string Morir(int nacimiento, int fallecimiento, string causa)
```
Comment (MauroDecodificado) [inline Colibri.cs]: also method start with lower case "morir"

File: Colibri.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            int calculo;
+            calculo = nacimiento - fallecimiento;
+            Console.WriteLine("La fecha de su muerte es: " + fallecimiento, "vivio "+ calculo, "años y la causa de la muerte es"+ causa);
+            string m=Console.ReadLine();
+            return m;
+        }  
+        //se retorna el mismo string que recibimos
+        public string Reencarna(string revivio)
```
Comment (MauroDecodificado) [inline Colibri.cs]: Method starts with lower case

File: Muerte.cs [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace EjercicioDos
+{
+    internal interface Muerte
+    { 
+        //Se crea crean compaortamientos que reciban y devuelvan parametros
+     string Morir(int nacimiento, int fallecimiento, string causa);
```
Comment (MauroDecodificado) [inline Muerte.cs]: Methods start with lower case

File: Vaca.cs [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+              basenumero  = "par";
+            }
+            else
+            {
+              basenumero = "impar";
+            }
+
+            Console.WriteLine("La fecha de su muerte " +fallecimiento, "el numero es " + basenumero, "vivio"+ calculo, "años y murio por "+ causa);
+            
+            return Console.ReadLine();
```
Comment (MauroDecodificado) [inline Vaca.cs]: Please apply the same suggestion for this for returning the message

**Review by MauroDecodificado — CHANGES_REQUESTED** [review]
Check comments

File: Colibri.cs (hunk 12 above)
Comment (RonalAlexis2) [reply Colibri.cs]: has been corrected

File: Colibri.cs (hunk 13 above)
Comment (RonalAlexis2) [reply Colibri.cs]: has been corrected

File: Muerte.cs (hunk 14 above)
Comment (RonalAlexis2) [reply Muerte.cs]: has been corrected

File: Colibri.cs (hunk 4 above)
Comment (RonalAlexis2) [reply Colibri.cs]: has been corrected

File: Vaca.cs (hunk 15 above)
Comment (RonalAlexis2) [reply Vaca.cs]: has been corrected

File: Colibri.cs [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        public void volar()
+        {
+            Console.WriteLine("el colibri vuela");
+        }
+
+        //se reciben los parametros de muerte y se calcula cuanto vivio
+        public string morir(int nacimiento, int fallecimiento, string causa)
+        {
+            int calculo;
+            calculo = nacimiento - fallecimiento;
```
Comment (MauroDecodificado) [inline Colibri.cs]: This is not the right approach to calculate age

File: Colibri.cs [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+            Console.WriteLine("La fecha de su muerte es: " + fallecimiento.ToString(), "vivio "+ calculo.ToString(), "años y la causa de la muerte es"+ causa);
+            
+            return Console.ReadLine();
```
Comment (MauroDecodificado) [inline Colibri.cs]: This does not make sense, you do not need to have this and return it. do this

```
	public string morir(int nacimiento, int fallecimiento, string causa)
        {
            int calculo;
            calculo = nacimiento - fallecimiento;
           return "La fecha de su muerte es: " + fallecimiento.ToString()+ " vivio "+ calculo.ToString()+ " años y la causa de la muerte es "+ causa
       
	}
```

and  the call it and print it from the program

File: Vaca.cs (hunk 15 above)
Comment (MauroDecodificado) [reply Vaca.cs]: Check the comment I just did

**Review by MauroDecodificado — CHANGES_REQUESTED** [review]

File: Vaca.cs (hunk 15 above)
Comment (RonalAlexis2) [reply Vaca.cs]: has been corrected

File: Colibri.cs (hunk 17 above)
Comment (RonalAlexis2) [reply Colibri.cs]: has been corrected

File: Colibri.cs (hunk 16 above)
Comment (RonalAlexis2) [reply Colibri.cs]: has been corrected

File: Program.cs [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+using System.Threading.Tasks;
+
+namespace EjercicioDos
+{
+    internal class Program
+    {
+        static void Main(string[] args)
+        {
+            Colibri ocolibri = new Colibri();
+            ocolibri.Morir(2022,2023,"hambre");
```
Comment (MauroDecodificado) [inline Program.cs]: Please here use 

 Console.WriteLine(ocolibri.Morir(2022,2023,"hambre");


in order to show the message in the console

File: Program.cs [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+            ocolibri.Reencarna("wuju");
```
Comment (MauroDecodificado) [inline Program.cs]: Please here use 

 Console.WriteLine(ocolibri.reencarna("wuju");


in order to show the message in the console

File: Program.cs [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+            ocolibri.atacar();
+            ocolibri.volar();
+
+            Vaca ovaca = new Vaca();
+            ovaca.Morir(2015,2023,"devorada");
```
Comment (MauroDecodificado) [inline Program.cs]: Same for those methods

**Review by MauroDecodificado — COMMENTED** [review]
Just minor comments, fix them and we are good to go

File: Program.cs (hunk 18 above)
Comment (RonalAlexis2) [reply Program.cs]: has been corrected

File: Program.cs (hunk 19 above)
Comment (RonalAlexis2) [reply Program.cs]: has been corrected

File: Program.cs (hunk 20 above)
Comment (RonalAlexis2) [reply Program.cs]: has been corrected

**Review by MauroDecodificado — APPROVED** [review]
PR Looks good now, and can be merged 😄

**Pull request merged**